tsdb/head_read.go

```diff
@@ -511,6 +550,80 @@ func (o mergedOOOChunks) NumSamples() int {
 // Compact TODO(jesus.vazquez) Clarify why this method wont be called
 func (o mergedOOOChunks) Compact() {}
 
+var _ chunkenc.Chunk = &boundedChunk{}
+
+// boundedChunk is an implementation of chunkenc.Chunk that uses a
+// boundedIterator that only iterates through samples which timestamps are
+// >= minT and <= maxT
+type boundedChunk struct {
+	chunkenc.Chunk
+	minT int64
+	maxT int64
+}
+
+func (b boundedChunk) Bytes() []byte {
+	xor := chunkenc.XORChunk{}
+	a, _ := xor.Appender()
+	it := b.Iterator(nil)
+	for it.Next() {
+		t, v := it.At()
+		a.Append(t, v)
+	}
+	return xor.Bytes()
+}
+
+func (b boundedChunk) Iterator(iterator chunkenc.Iterator) chunkenc.Iterator {
+	it := b.Chunk.Iterator(iterator)
+	if it == nil {
+		panic("iterator shouldn't be nil")
+	}
+	return boundedIterator{it, b.minT, b.maxT}
+}
+
+var _ chunkenc.Iterator = &boundedIterator{}
+
+// boundedIterator is an implementation of Iterator that takes two timestamp
+// delimitators minT and maxT and only iterates through samples within
+// those timestamps
+type boundedIterator struct {
+	chunkenc.Iterator
+	minT int64
+	maxT int64
+}
+
+// Next the first time its called it will advance as many positions as necessary
+// until its able to find a sample within the bounds minT and maxT.
+// If there are samples within bounds it will advance one by one amongst them.
+// If there are no samples within bounds it will return false.
+func (b boundedIterator) Next() bool {
+	for b.Iterator.Next() {
+		t, _ := b.Iterator.At()
+		if t < b.minT {
+			continue
+		} else if t > b.maxT {
+			return false
+		}
+		break
+	}
+	return true
+}
+
+func (b boundedIterator) Seek(t int64) bool {
+	if t < b.minT || t > b.maxT {
+		return false
+	}
+	return b.Iterator.Seek(t)
+}
```

Comment:
Seek is like a pointer in the iterator that only moves forward. If seek is called for `t`, it seeks until it finds a sample that is `>= t`

So when `t < minT`, the first Seek'ed sample must be after minT, so we seek for minT instead and make sure the seeked sample if before maxT.

Next if `t > maxT`, it means that it permanently moves forward the pointer there, so any subsequent Seek() or Next() call must return false. Hence we move forward the pointer.

```suggestion
func (b boundedIterator) Seek(t int64) bool {
	if t < b.minT {
		// We must seek at least up to b.minT if it is asked for something before that.
		ok := b.Iterator.Seek(b.minT)
		if !ok {
			return false
		}
		t, _ := b.Iterator.At()
		return t <= b.maxT 
	}
	if t > b.maxT {
		// We seek anyway so that the subsequent Next() calls will also return false.
		b.Iterator.Seek(t)
		return false
	}
	return b.Iterator.Seek(t)
}
```